### PR TITLE
Add support for drug-drug interactions with order-select hook.

### PR DESCRIPTION
**Pull request by plarocque4**

### CODE REVIEW

**Comment by plarocque4** [comment]
Test with order-select branches for CRD, CDS-Library, crd-request-generator and test-ehr.

**Comment by yuliumitre** [comment]
Hi, Pat, I am running CRD, request-generator, test-ehr in branch order-select.  For some reason, select medication methotrexate did trigger order-select, but the card is empty. But I thought you got warning back during your demo. Did I miss anything? 
Screen Shot 2021-07-16 at 3 07 25 PM
Screen Shot 2021-07-16 at 3 07 32 PM

**Comment by plarocque4** [comment]
> Hi, Pat, I am running CRD, request-generator, test-ehr in branch order-select. For some reason, select medication methotrexate did trigger order-select, but the card is empty. But I thought you got warning back during your demo. Did I miss anything?
> Screen Shot 2021-07-16 at 3 07 25 PM
> Screen Shot 2021-07-16 at 3 07 32 PM

Are you running with all four branches and the data loaded into the test-ehr? You need CRD, crd-request-generator, CDS-Library, and test-ehr.

**Comment by yuliumitre** [comment]
> > Hi, Pat, I am running CRD, request-generator, test-ehr in branch order-select. For some reason, select medication methotrexate did trigger order-select, but the card is empty. But I thought you got warning back during your demo. Did I miss anything?
> 
> Are you running with all four branches and the data loaded into the test-ehr? You need CRD, crd-request-generator, CDS-Library, and test-ehr.

Ah, I missed CDS-Library repo. It works now. Thanks Pat!

File: resources/src/main/java/org/cdshooks/CdsResponse.java [hunk 1]
```diff
@@ -11,7 +11,7 @@
    * them). The EHR decides how to display cards, but we recommend displaying suggestions using
    * buttons, and links using underlined text. REQUIRED
    */
-  private List<Card> cards = null;
+  private List<Card> cards = new ArrayList<Card>();
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply resources/src/main/java/org/cdshooks/CdsResponse.java:14]: Thanks, will remove it.

File: resources/src/main/java/org/cdshooks/CoverageRequirements.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  private String infoLink;
+  private String questionnaireOrderUri;
+  private String questionnaireFaceToFaceUri;
+  private String questionnaireLabUri;
+  private String questionnaireProgressNoteUri;
+  private String questionnairePARequestUri;
+  private String questionnairePlanOfCareUri;
+  private String questionnaireDispenseUri;
+  private String requestId;
+  private Boolean priorAuthRequired;
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply resources/src/main/java/org/cdshooks/CoverageRequirements.java:16]: Thanks, updating them to all be the primitive type.

File: resources/src/main/java/org/cdshooks/CoverageRequirements.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  public boolean getPriorAuthRequired() {
+    return priorAuthRequired;
+  }
+
+  public CoverageRequirements setPriorAuthRequired(boolean priorAuthRequired) {
+    this.priorAuthRequired = priorAuthRequired;
+    return this;
+  }
+
+  public boolean getDocumentationRequired() {
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply resources/src/main/java/org/cdshooks/CoverageRequirements.java:132]: Thanks, will update it.

File: server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSelectService.java [hunk 4]
```diff
@@ -45,14 +46,64 @@
...
+
+    AlternativeTherapy alternativeTherapy = new AlternativeTherapy();
+    alternativeTherapy.setApplies(false);
+    results.setAlternativeTherapy(alternativeTherapy);
+
+    DrugInteraction drugInteraction = new DrugInteraction();
+    drugInteraction.setApplies(false);
+    try {
+      if (evaluateStatement("DRUG_INTERACTION", context) != null) {
+        Object b = evaluateStatement("DRUG_INTERACTION", context);
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSelectService.java:78]: Will remove it.

File: server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSignService.java [hunk 5]
```diff
@@ -55,4 +61,119 @@
...
+
+    boolean isNotMedicationDispense = true;
+    if (evaluateStatement("RESULT_dispense", context) != null) {
+      results.setRequest((IBaseResource) evaluateStatement("RESULT_dispense", context));
+      coverageRequirements.setRequestId(JSONObject.escape(fhirComponents.getFhirContext().newJsonParser()
+          .encodeResourceToString(results.getRequest())));
+      isNotMedicationDispense = false;
+    }
+
+    if (evaluateStatement("RESULT_QuestionnaireOrderUri", context) != null && isNotMedicationDispense) {
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSignService.java:98]: Good idea, I refactored out the variable as well.

File: server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/CdsService.java [hunk 6]
```diff
@@ -158,205 +158,114 @@ public CdsResponse handleRequest(@Valid @RequestBody requestTypeT request, URL a
...
     // CQL Executed
     requestLog.advanceTimeline(requestService);
 
-    if (!foundApplicableRule) {
-      String msg = "No documentation rules found";
-      logger.warn(msg + "; summary card sent to client");
-      response.addCard(CardBuilder.summaryCard(msg));
-    }
+    if (errorCardOnEmpty) {
+      if (!foundApplicableRule) {
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/CdsService.java:211]: Good catch, this doesn't work as desired. I wanted errorCardOnEmpty to be false on a order-select request. I changed it to work this way now.

File: server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSelectService.java [hunk 7]
```diff
@@ -45,14 +46,64 @@
...
+        drugInteraction.setApplies((Boolean) evaluateStatement("DRUG_INTERACTION", context));
+
+        if (drugInteraction.getApplies()) {
+          drugInteraction.setSummary("WARNING! Drug Interaction Found!");
+          String detail = "Drug ";
+
+          if (evaluateStatement("REQUESTED_DRUG_CODE", context) != null) {
+            Object c = evaluateStatement("REQUESTED_DRUG_CODE", context);
+            ArrayList<Coding> clist = (ArrayList<Coding>) c;
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSelectService.java:87]: Done

File: server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSelectService.java [hunk 8]
```diff
@@ -45,14 +46,64 @@
...
+            Coding code = clist.get(0);
+            detail = detail + " " + code.getDisplay() + " (" + code.getCode() + ") has a dangerous drug/drug interaction with medication patient is already taking: ";
+          }
+          if (evaluateStatement("STATEMENT_DRUG_CODE", context) != null) {
+            Object c = evaluateStatement("STATEMENT_DRUG_CODE", context);
+            ArrayList<Coding> clist = (ArrayList<Coding>) c;
```
[the remark this answers is not in the document]
Comment (plarocque4) [reply server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSelectService.java:93]: and done

File: server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSignService.java [hunk 9]
```diff
@@ -87,65 +87,65 @@ protected CqlResultsForCard executeCqlAndGetRelevantResults(Context context, Str
...
+      } catch (Exception e) {
+        logger.info("-- No PA Request questionnaire defined");
       }
-    } catch (Exception e) {
-      logger.info("-- No PA Request questionnaire defined");
-    }
 
-    // only display the dispense form for MedicationDispense request
-    try {
-      if (evaluateStatement("RESULT_QuestionnaireDispenseUri", context) != null && !isNotMedicationDispense) {
```
Comment (KeeyanGhoreshi) [inline server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSignService.java]: this was originally run if `!isNotMedicationDispense`, unlike the others which are just `isNotMedicationDispense`.  Should this be moved out of the else block?

Comment (plarocque4) [reply server/src/main/java/org/hl7/davinci/endpoint/cdshooks/services/crd/r4/OrderSignService.java]: Good catch, that would have broken MedicationDispense. I have pushed up a fix.

**Review by KeeyanGhoreshi — APPROVED** [review]

**Review by KeeyanGhoreshi — APPROVED** [review]

**Pull request merged**